wallet-new/src/Cardano/Wallet/API/Types.hs

```diff
@@ -1,23 +1,161 @@
+    -> (Proxy api' -> SI.Context ctx -> SI.Delayed env (Server api') -> SI.Router env)
+inRouteServer routing f = \_ ctx delayed -> routing Proxy ctx (fmap f delayed)
+
+-- | `ApiHasArg*` types help to reflect on names and types of arguments of API methods,
+-- specified by `Capture`s, `QueryParam`s, etc. These types are used to construct more
+-- complex useful types such as `WithDefaultApiArg`.
+class ApiHasArgClass apiType a where
+    -- | For arguments-specifiers of API, get argument type.
+    type ApiArg (apiType :: * -> *) a :: *
+    type ApiArg apiType a = a
+
+    -- | Name of argument.
+    -- E.g. name of argument specified by @Capture "nyan"@ is /nyan/.
+    apiArgName :: Proxy (apiType a) -> String
+    default apiArgName
+        :: forall n someApiType. (KnownSymbol n, someApiType n ~ apiType)
+        => Proxy (someApiType n a) -> String
+    apiArgName _ = symbolVal (Proxy @n)
+
+type ApiHasArgInvariant apiType a res =
+    Server (apiType a :> res) ~ (ApiArg apiType a -> Server res)
+
+type ApiHasArg apiType a res =
+    ( ApiHasArgClass apiType a
+    , ApiHasArgInvariant apiType a res
+    )
+
+instance KnownSymbol s => ApiHasArgClass (Capture s) a
+instance KnownSymbol s => ApiHasArgClass (QueryParam s) a where
+    type ApiArg (QueryParam s) a = Maybe a
+instance KnownSymbol s => ApiHasArgClass (QueryParams s) a where
+    type ApiArg (QueryParams s) a = [a]
+instance KnownSymbol s => ApiHasArgClass (Header s) a where
+    type ApiArg (Header s) a = Maybe a
+instance ApiHasArgClass (ReqBody ct) a where
+    apiArgName _ = "request body"
+
+-- | `Unmaybe` and `UnmaybeArg` type families are necessary to
+-- get `a` from `Maybe a` in a way which doesn't make GHC mad.
+-- Used in definition of `ApiHasArgClass` and `HasServer` instances for
+-- `WithDefaultApiArg`
+type family Unmaybe a where
+    Unmaybe (Maybe a) = a
+
+type family UnmaybeArg a where
+    UnmaybeArg (Maybe a -> b) = a -> b
+
+--
+-- Specifying API method parameters with default values
+--
+
+-- | `WithDefaultApiArg` type is used to specify API parameters,
+-- which are optional but have default values. It can be applied to every
+-- API specifier which yields `Maybe a` as argument type.
+data WithDefaultApiArg (argType :: * -> *) a
+
+instance (ApiHasArgClass apiType a, ApiArg apiType a ~ Maybe b) =>
+         ApiHasArgClass (WithDefaultApiArg apiType) a where
+    type ApiArg (WithDefaultApiArg apiType) a = Unmaybe (ApiArg apiType a)
+    apiArgName _ = apiArgName (Proxy @(apiType a))
+
+instance ( HasServer (apiType a :> res) ctx
+         , Server (apiType a :> res) ~ (Maybe c -> d)
+         , Default c
+         ) =>
+         HasServer (WithDefaultApiArg apiType a :> res) ctx where
+    type ServerT (WithDefaultApiArg apiType a :> res) m =
+        UnmaybeArg (ServerT (apiType a :> res) m)
+    route =
+        inRouteServer @(apiType a :> res) route $
+        \f a -> f $ fromMaybe def a
+
+instance HasSwagger (apiType a :> res) =>
```

Comment:
Minor nitpick: If we move the `HasSwagger` definition inside `V1.Swagger` we don't need to depend on `servant-swagger`. It's something I'm trying to achieve as part of the design: there should be separation of concerns between the logic & the documentation.

What do you think, @flyingleafe ?

Reply:
It's sensible.